mtags/src/main/scala/scala/meta/internal/mtags/JavaToplevelMtags.scala

```diff
@@ -0,0 +1,272 @@
+package scala.meta.internal.mtags
+
+import scala.annotation.tailrec
+
+import scala.meta.dialects
+import scala.meta.inputs.Input
+import scala.meta.inputs.Position
+import scala.meta.internal.semanticdb.Language
+import scala.meta.internal.semanticdb.SymbolInformation
+import scala.meta.internal.tokenizers.CharArrayReader
+import scala.meta.internal.tokenizers.Chars._
+import scala.meta.internal.tokenizers.Reporter
+
+class JavaToplevelMtags(val input: Input.VirtualFile) extends MtagsIndexer {
+
+  import JavaToplevelMtags._
+
+  val reporter: Reporter = Reporter(input)
+  val reader: CharArrayReader =
+    new CharArrayReader(input, dialects.Scala213, reporter)
+
+  override def language: Language = Language.JAVA
+
+  override def indexRoot(): Unit = {
+    if (!input.path.endsWith("module-info.java")) {
+      reader.nextRawChar()
+      loop
+    }
+  }
+
+  private def loop: Unit = {
+    val token = fetchToken
+    token match {
+      case Token.EOF =>
+      case Token.Package =>
+        val paths = readPaths
+        paths.foreach { path => pkg(path.value, path.pos) }
+        loop
+      case Token.Class | Token.Interface | _: Token.Enum | _: Token.Record =>
+        fetchToken match {
+          case Token.Word(v, pos) =>
+            val kind = token match {
+              case Token.Interface => SymbolInformation.Kind.INTERFACE
+              case _ => SymbolInformation.Kind.CLASS
+            }
+            withOwner(currentOwner)(tpe(v, pos, kind, 0))
+            skipBody
+            loop
+          case _ =>
+            loop
+        }
+      case _ =>
+        loop
+    }
+  }
+
+  @tailrec
+  private def skipMultilineComment(prevStar: Boolean = false): Unit = {
+    reader.nextChar()
+    if (prevStar) {
+      if (reader.ch == '/') reader.nextChar()
+      else skipMultilineComment(prevStar = reader.ch == '*')
+    } else skipMultilineComment(prevStar = reader.ch == '*')
+  }
+
+  private def fetchToken: Token = {
+
+    @tailrec
+    def quotedLiteral(quote: Char): Token = {
+      reader.nextChar()
+      reader.ch match {
+        case `quote` => Token.Literal
+        case '\\' =>
+          reader.nextChar()
+          quotedLiteral(quote)
+        case _ => quotedLiteral(quote)
+      }
+    }
+
+    @tailrec
+    def kwOrIdent(start: Int, builder: StringBuilder): Token = {
+      val ch = reader.ch
+      if (Character.isJavaIdentifierPart(ch)) {
+        reader.nextChar()
+        kwOrIdent(start, builder.append(ch))
+      } else if (builder.isEmpty) {
+        throw new Exception(
```

Comment:
When the exception is thrown are we still able to find the toplevel? Do we revert to the previous logic?